Sources/AsyncAlgorithms/zip/Zip2StateMachine.swift

```diff
@@ -0,0 +1,519 @@
+where Base1: Sendable, Base1.Element: Sendable, Base2: Sendable, Base2.Element: Sendable {
+  private let stateMachine = Zip2StateMachine<Base1.Element, Base2.Element>()
+
+  init(_ base1: Base1, _ base2: Base2) {
+    let task = Task { [stateMachine] in
+      await withTaskGroup(of: Void.self) { group in
+        group.addTask {
+          var base1Iterator = base1.makeAsyncIterator()
+
+          do {
+          loop: while true {
+            await withUnsafeContinuation{ continuation in
+              stateMachine.newLoopFromBase1(suspendedBase: continuation)
+            }
+
+            guard let element1 = try await base1Iterator.next() else {
+              break loop
+            }
+
+            stateMachine.base1HasProducedElement(element: element1)
+          }
+          } catch {
+            stateMachine.base1HasProducedFailure(error: error)
+          }
+
+          stateMachine.baseIsFinished()
+        }
+
+        group.addTask {
+          var base2Iterator = base2.makeAsyncIterator()
+
+          do {
+          loop: while true {
+            await withUnsafeContinuation{ continuation in
+              stateMachine.newLoopFromBase2(suspendedBase: continuation)
+            }
+
+            guard let element2 = try await base2Iterator.next() else {
+              break loop
+            }
+
+            stateMachine.base2HasProducedElement(element: element2)
+          }
+          } catch {
+            stateMachine.base2HasProducedFailure(error: error)
+          }
+          
+          stateMachine.baseIsFinished()
+        }
+      }
+    }
+
+    self.stateMachine.registerTask(task: task)
+  }
+
+  func next() async rethrows -> (Base1.Element, Base2.Element)? {
+    try await withTaskCancellationHandler { [stateMachine] in
+      stateMachine.rootTaskIsCancelled()
+    } operation: { [stateMachine] in
+      let results = await withUnsafeContinuation { continuation in
+        stateMachine.newDemandFromConsumer(suspendedDemand: continuation)
+      }
+
+      guard let results else {
+        return nil
+      }
+
+      stateMachine.demandIsFulfilled()
+
+      return try (results.0._rethrowGet(), results.1._rethrowGet())
+    }
+  }
+}
+
+final class Zip2StateMachine<Element1, Element2>: Sendable
+where Element1: Sendable, Element2: Sendable {
+  let state = ManagedCriticalState<State>(.initial(task: nil))
+
+  enum State: Sendable {
+    case initial(task: Task<Void, Never>?)
+    case awaitingDemandFromConsumer(
+      task: Task<Void, Never>?,
+      suspendedBases: [UnsafeContinuation<Void, Never>]
+    )
+    case awaitingBaseResults(
+      task: Task<Void, Never>?,
+      result1: Result<Element1, Error>?,
+      result2: Result<Element2, Error>?,
+      suspendedBases: [UnsafeContinuation<Void, Never>],
+      suspendedDemand: UnsafeContinuation<(Result<Element1, Error>, Result<Element2, Error>)?, Never>
+    )
+    case basesResultsAreReady(
```

Comment:
I think we can skip this state entirely. If we are in `awaitingBaseResults` and we get the last base element we can transition back to `awaitingDemandFromConsumer`. The only thing that need to change is being able to buffer values in `awaitingDemandFromConsumer`

Reply:
Done.